### PR TITLE
Support RocksDB database

### PR DESCRIPTION
**Pull request by gumb0**
Based on https://github.com/ethereum/cpp-ethereum/pull/4697
TODO:
- ~~[ ] Select either RocksDB or LevelDB in cmake files  based on `DB` build option~~ postponed
- ~~[ ] Build `RocksDB.*` only if `DB` is `ROCKSDB`~~ postponed
- [x] Fix RocksDB build for all required platforms

cc @isaachier

### CODE REVIEW

**Comment by gumb0** [comment]
Currently RocksDB build fails on Windows and on GCC6

**Comment by gumb0** [comment]
Error handling needs to be reworked here according to the changes in https://github.com/ethereum/cpp-ethereum/pull/4876

**Comment by gumb0** [comment]
Rebased.

**Comment by gumb0** [comment]
Current status: need to update PR to get RocksDB version 5.14 from Hunter, then continue with items from top comment

**Comment by halfalicious** [comment]
> Current status: need to update PR to get RocksDB version 5.14 from Hunter, then continue with items from top comment

Can you please clarify what you mean by this (sorry if it's obvious, I'm new to Hunter)? The following in `libdevcore\CMakeLists.txt` doesn't specify a version number so it should presumably just pull the latest version:

```
# if(DB STREQUAL "ROCKSDB")
    hunter_add_package(rocksdb)
    find_package(RocksDB CONFIG REQUIRED)
# endif()

```
Note that the latest version (according to Hunter's `default.cmake` file) is 5.14.2: https://github.com/ruslo/hunter/blob/cb14145ae215209718366cbc4079a7120d82e2c3/cmake/configs/default.cmake#L345

**Comment by gumb0** [comment]
@halfalicious That comment was written before 5.14 was released, so now we need just to get the latest from Hunter. I think just rebasing this PR should theoretically be enough to fix windows build.

So `RocksDB` implemenation should basically be ready here, all we need then is to figure out how to handle switching between RocksDB & LevelDB (at compile time, or maybe just build both for starters)

**Comment by halfalicious** [comment]
This still builds 5.8.6 after rebasing for some reason, I'll take a closer look later this week and see if I can figure out why.

**Comment by halfalicious** [comment]
Rebased

**Comment by gumb0** [comment]
Probably we need to update Hunter version (older hunter package list has older RocksDB only)

Try replacing this with the latest version from Hunter repo
https://github.com/ethereum/aleth/blob/561485d36d2690f5c74b4cfd743ab572e6ab80b5/CMakeLists.txt#L40-L45

**Comment by halfalicious** [comment]
> I think this is just about ready for review. The changes (which enable RocksDB to be selected at runtime in both Aleth and TestEth) build clean on Windows 10 x64 and I've done some basic sanity-testing (run the blockchain basic insert test case and start a full sync). I'm going to perform some more extensive testing before requesting a review (run through all of the tests locally on Windows and also build / run through some tests on Linux).

I've run through all of the tests on Windows and Linux (latest version of Ubuntu i.e. 18.04) using the RocksDB backend and they all pass. 

I'd like to also start syncing the blockchain using the new backend, I'll do that later this week. Once I've verified that works successfully I'll request eyes on this review.

**Comment by halfalicious** [comment]
> > I think this is just about ready for review. The changes (which enable RocksDB to be selected at runtime in both Aleth and TestEth) build clean on Windows 10 x64 and I've done some basic sanity-testing (run the blockchain basic insert test case and start a full sync). I'm going to perform some more extensive testing before requesting a review (run through all of the tests locally on Windows and also build / run through some tests on Linux).
> 
> I've run through all of the tests on Windows and Linux (latest version of Ubuntu i.e. 18.04) using the RocksDB backend and they all pass.
> 
> I'd like to also start syncing the blockchain using the new backend, I'll do that later this week. Once I've verified that works successfully this change will be ready for review.

I've successfully sync'd a few thousand Ropsten blocks on both Windows and Linux so I think these changes are ready for review. One note - the tests took approximately 70% longer to run when using RocksDB so it looks like RocksDB has worse performance, at least for the read/write patterns used in the tests. I didn't quiet my machine before running the tests so these aren't scientific performance comparisons but I didn't have anything intensive running in the background and I didn't use my machine while the tests were being executed so I think the results are somewhat accurate. Regarding hard numbers, I ran 2 iterations of the tests using LevelDB (results: 982 and 1019) and 2 iterations using RocksDB (results: 1738 and 1769). 

I'm not sure if this significant perf delta is expected? I did a quick online search of RocksDB vs LevelDB perf comparisons and I didn't see anything indicating that RocksDB was expected to perform worse. 

I'll start taking a look at the runtime deltas for specific tests to see if I can identify any patterns.

**Comment by halfalicious** [comment]
Note that I chose to make RocksDB selectable at runtime in Aleth and TestEth (via the --db argument in both cases) since that was more straightforward and makes testing each implementation easier. I think once we've verified that RocksDB works well we can either remove LevelDB entirely or make the option compile time. 

Also, note that if one syncs with LevelDB and then tries to open the DB with Rocks DB (or vice-versa), Aleth will exit with a corrupt database error. This is expected and I think this is acceptable given that we only expect to support both databases temporarily.

**Comment by halfalicious** [comment]
I'm a little confused about the AppVeyor build failure...I skimmed through the log and saw only warnings, not errors. Are there issues with AppVeyor builds being flaky?

**Comment by chfast** [comment]
I did Hunter upgrade alone in https://github.com/ethereum/aleth/pull/5339. Let's merge that one first and the rebase this one. The Hunter cache should be up to date now.

**Comment by halfalicious** [comment]
> I did Hunter upgrade alone in #5339. Let's merge that one first and the rebase this one. The Hunter cache should be up to date now.

Thanks, I'll merge that change and then rebase this one

**Comment by halfalicious** [comment]
Looks like I somehow screwed up the commit history, I'll investigate what happened and see if I can fix it.

File: libdevcore/DBFactory.h [hunk 1]
```diff
@@ -41,6 +42,7 @@ boost::program_options::options_description databaseProgramOptions(
     unsigned _lineLength = boost::program_options::options_description::m_default_line_length);
 
 bool isMemoryDB();
+bool isDiskDatabase();
```
[the remark this answers is not in the document]
Comment (halfalicious) [reply libdevcore/DBFactory.h]: I removed the isMemoryDB function since we don't need both an isDiskDatabase and an isMemoryDatabase function

**Comment by halfalicious** [comment]
Rebased and good to go, if no one has any objections I'll merge.

**Comment by halfalicious** [comment]
> I can't approve this because I was the original author @chfast

I was able to approve since I wasn't the author

**Pull request merged**